### DISCUSSION TITLE
Support for FETCH after PIPE

### DISCUSSION THREAD

**Discussion by vishalratnam-cliq-scm** [vesoft-inc/nebula#4188] [Ideas]
Discussion around the nebula forums on with @wey-gu  : [https://discuss.nebula-graph.io/t/retrieve-all-edges-with-all-ranks-using-fetch-query/1195/5](https://discuss.nebula-graph.io/t/retrieve-all-edges-with-all-ranks-using-fetch-query/1195/5)
I’m using a composite query involving GO and FETCH.
Query:
`GO FROM “v1_id” over edgeName1 yield $$.v2.id as alliasId | FETCH PROP ON edgeName2 $-.alliasId → “v3_vid” YIELD properties(edge);`
Example at: [https://docs.nebula-graph.io/2.0/3.ngql-guide/7.general-query-statements/4.fetch/#use_fetch_in_composite_queries](https://docs.nebula-graph.io/2.0/3.ngql-guide/7.general-query-statements/4.fetch/#use_fetch_in_composite_queries)

Required Traversal: v1 → v2 → v3

I need to traverse from a known v1_vid to a known v3_vid. The output of v2_vid should be used as an input to the FETCH query (as the number of edges of type edgeName2 is around 2 Billion and FETCH performs faster).
With the above query, I get a syntax error at `’ YIELD '. Looks like the problem is while providing the v3_vid in the query, which can’t be harcoded/provided. But since the v3_vid needs to be provided by us, this query is not working.

Is there a workaround to provide the v3_vid by the user? Or is it mandatory for v3_vid to be an output of the GO query.

Feature Required: Add support for `FETCH` query after `PIPE ( | )` so that the destination vertex can be a user defined string/ harcoded value.

**Comment by akhilravuri-tul-scm** [comment]
Yes, This is a good feature. We are also looking out for this.

**Reply by jamieliu1023** [reply]
@chvr203 Thanks for raising the issue here. Quote @wey-gu:

Could you try use GO instead of FETCH after the pipe?

```
GO FROM "player100" OVER follow yield $$.player.name, id($$) as alliasId | \
GO from $-.alliasId OVER follow WHERE id($$) == "player102" YIELD properties(edge)
```
FETCH doesn't support this pattern yet. If the performance cannot fulfill your requirement, please help raise an issue for adding support of FETCH after pipe :)

**Reply by vishalratnam-cliq-scm** [reply]
Hi @jamieliu1023, 
Yes, this works as expected. But we're developing an application that contains around 2 Billion edges and 100k vertices.
After looking at the docs and trying some examples, we found that `FETCH` is the fastest performing query. We see that we get response in as low as 5ms with `FETCH` query on our data. So, ideally we would need the `FETCH` query to support `PIPE(|)` for faster retrieval.
The given query (GO pipe with GO) works, but takes around 62seconds!

**Reply by jamieliu1023** [reply]
@chvr203 Thanks for the feedback. I have created an issue from this discussion as a feature request.

**Reply by xiajingchun** [reply]
@chvr203 could you try below:
GO FROM “v1_id” over edgeName1 yield $$.v2.id as alliasId, "v3_vid" as v3 | FETCH PROP ON edgeName2 $-.alliasId → $-.v3 YIELD properties(edge);
It kinda workaround the limitation by passing the hardcoded value also through the pipe.

**Reply by vishalratnam-cliq-scm** [reply]
@xiajingchun thank you, this works as expected!
However, the actual syntax of the `FETCH` query allows us to fetch multiple edges of different ranks by comma separation. 

When I try to run the following query, it works:
`GO FROM “v1_id” over edgeName1 yield $$.v2.id as alliasId, "v3_vid" as v3 | FETCH PROP ON edgeName2 $-.alliasId → $-.v3 @1 YIELD properties(edge);`

But when I run this query to fetch multiple edges with different ranks, I get a syntax error.
`GO FROM “v1_id” over edgeName1 yield $$.v2.id as alliasId, "v3_vid" as v3 | FETCH PROP ON edgeName2 $-.alliasId → $-.v3 @1,  $-.alliasId → $-.v3 @2 YIELD properties(edge);`

Notice the `, $-.alliasId → $-.v3 @2` in the second query to fetch the edges that have rank 2 between the given source and destination vertices. 
 
Is there any way to make this work?

**Reply by xiajingchun** [reply]
Unfortunately this is not supported. If you don't really need to put the two edges with different ranks into one row, you may try to UNION two fetch statements. In my testing, I couldn't figure it out with pipe, but it's possible to use variables.

**Reply by wey-gu** [reply]
To know that the case both peers of edge are variables is supported helped a lot @xiajingchun.

Reported https://github.com/vesoft-inc/nebula/issues/4222 on not being able to support multiple edges being fetched with variable peers.

As @xiajingchun 's solution, `UNION` with proper brackets and variable(pipe case not parsed as expected now, neither) now can mitigate the case.
```
(root@nebula) [basketballplayer]> $var = go from "player100" over follow yield id($$) as alliasId, "player102" as v3 ; (FETCH PROP ON follow $var.alliasId -> $var.v3 YIELD properties(edge), edge as e UNION FETCH PROP ON follow $var.alliasId -> $var.v3@1 YIELD properties(edge), edge as e);
+------------------+----------------------------------------------------+
| properties(EDGE) | e                                                  |
+------------------+----------------------------------------------------+
| {degree: 90}     | [:follow "player101"->"player102" @0 {degree: 90}] |
| {degree: 10}     | [:follow "player101"->"player102" @1 {degree: 10}] |
+------------------+----------------------------------------------------+
```

**Reply by vishalratnam-cliq-scm** [reply]
Hey @wey-gu @xiajingchun, this works like a charm! Thanks a lot!

**Comment by wey-gu** [comment]
**Summary**

Issues for now:

1. https://github.com/vesoft-inc/nebula/issues/4222 FETCH PROP on edge with variable not supporting multiple edges as expected
2. https://github.com/vesoft-inc/nebula/issues/4191 FETCH PROP on edge with variable only support non or both peers to be variable, not the case when mixed
3. UNION cannot be used after PIPE, but could be mitigated with variable

Mitigation for now thanks to @xiajingchun :

- for 2. #4191, just put the non-variable peer as the variable, too
- for 1. #4222 , use UNION and for 3. , use $var = xxx instead of PIPE

```
(root@nebula) [basketballplayer]> $var = go from "player100" over follow yield id($$) as alliasId, "player102" as v3 ; (FETCH PROP ON follow $var.alliasId -> $var.v3 YIELD properties(edge), edge as e UNION FETCH PROP ON follow $var.alliasId -> $var.v3@1 YIELD properties(edge), edge as e);
+------------------+----------------------------------------------------+
| properties(EDGE) | e                                                  |
+------------------+----------------------------------------------------+
| {degree: 90}     | [:follow "player101"->"player102" @0 {degree: 90}] |
| {degree: 10}     | [:follow "player101"->"player102" @1 {degree: 10}] |
+------------------+----------------------------------------------------+
```
